[LOOP-345] First pass of Nightscout remote data pull synchronization

darinkrauss opened the pull request:
@ps2 In order to expedite the feedback & update process, would you please do a first pass review over this entire PR (and the other two related PRs) with an eye towards whether this pull synchronization solution meets your thinking from an architectural and design standpoint. After you and I work through any needed changes, I'll add in Nate to do a review all of the code level details.

darinkrauss committed:
First pass of pull synchronization

- Add modificationCounter to relevant CoreData types
- Add new synchronization fields to DeletedCarbEntry
- Update CarbStore to initiate remote data synchronization when appropriate
- Add CarbRemoteDataQueryDelegate protocol support to CarbStore
- Add GlucoseRemoteDataQueryDelegate protocol support to GlucoseStore
- Update DoseStore to initiate remote data synchronization when appropriate
- Add DoseRemoteDataQueryDelegate protocol support to DoseStore
- Add modificationCounter support in NSManagedObjectContext extension
- Update RemoteDataService to support pull synchronization
- Add SettingStore to retain any settings changes while app is running
- Add StatusStore to retain any status updates while app is running
- Remove CarbStore.syncExternalDB
- Remote DoseStore.uploadPumpEventsIfNeeded
- Update MockService to support updated RemoteDataService

darinkrauss committed:
Update data model version

mpangburn committed:
Align basal schedule time zone with current calendar

ps2 commented on LoopKit/CarbKit/CarbStore.swift:
```diff
@@ -28,23 +28,11 @@ public protocol CarbStoreDelegate: class {
+    /// Ask the delegate to initiate remote data synchronization. Remote data synchronization is an asynchronous
+    /// process that pulls data from multiple local sources and synchronizes with any remote data services. The process
+    /// can take considerable time and no completion handler is provided.
+    func initiateRemoteDataSynchronization()
```
I think something like `func carbStoreDidUpdate(_ carbStore: CarbStore)` might be better here.

darinkrauss replied:
Done.

ps2 commented on LoopKit/CarbKit/DeletedCarbEntry.swift:
```diff
@@ -9,21 +9,29 @@ import Foundation
 
 public struct DeletedCarbEntry {
-    public let externalID: String
+    public let externalID: String?
```
Shouldn't this go away in favor of syncIdentifier?

darinkrauss replied:
Yes, but there is an intermediate time for DIY Loop when they were running an older version and now they are running a version that contains this code were we support both `externalID` and `syncIdentifier`. It is just supporting the post-upgrade time period of a few days while all of the "old" `externalID` data is still cached locally.

ps2 replied:
Can the role of external id migrate to syncIdentifier during the upgrade?

darinkrauss replied:
Hmm, not sure. I vaguely recall that the Nightscout endpoint for deleting via external ID versus syncIdentifer may be different. I'll look into it.

darinkrauss replied:
While the `syncIdentifier` is a new field on the `DeletedCarbObject` model, it is not on the `CachedCarbObject` model. To keep symmetry between the two models, I'd prefer to keep this as-is.  Eventually we can just delete all reference to `externalID` in both since it won't be used at all.

We cannot migrate `externalID` to `syncIdentifier` in `CachedCarbObject` because both fields are used for different purposes in existing installations of Loop. Specifically, the `externalID` field is used as the external Nightscout identifier (once the carb object is uploaded), while the `syncIdentifier` field is used to accurately update or delete any given carb object in HealthKit.

ps2 commented on LoopKit/RemoteDataService.swift:
```diff
@@ -7,30 +7,572 @@
+/**
+ A basic core data query anchor that uses a modification counter to track changes.
+ */
+public struct CoreDataQueryAnchor: QueryAnchor {
```
I don't think this belongs in RemoteDataService. It should be part of the API for the data stores.

darinkrauss replied:
Done.

ps2 reviewed (commented):
Good steps towards a pull architecture.  I think there should be a cleaner separation of DataStore api vs Service implementation. DataStores should define the anchors (as opaque types) and the remote data services are just clients of those stores (or maybe even just clients of a higher manager object in Loop that coordinates the interaction between the stores and the services).

darinkrauss committed:
Refactor pull synchronization to move main logic into remote data service manager

darinkrauss commented:
@ps2 Would you please review this latest revision of remote data service synchronization so we we can hopefully wrap this up this sprint?

Please review the PRs for the three repos (Loop, LoopKit, and NightscoutService) in their entirety (not just the latest commit). Also note that once we get the design in a good state, I'll rebase it on to the latest dev branch and submit a final set of PRs for review by both you and Nate.

Also, there are two issues that I'll fix when I return. First, there is a crashing bug (in code that should not crash), but it is either a lock or dispatch queue issue or perhaps a debugging issue. (I ran into this last, so haven't even tried a clean build.) Second, I need to add back a small bit of code to fully synchronize a remote data service when it is first added by the user to Loop.

Hopefully I have everything straightened out now. Thanks in advance!

ps2 commented on LoopKit/CarbKit/CachedCarbObject+CoreDataClass.swift:
```diff
@@ -50,6 +50,14 @@ class CachedCarbObject: NSManagedObject {
+
+    override func willSave() {
+        if isInserted || isUpdated {
+            setPrimitiveValue(managedObjectContext?.modificationCounter ?? 0, forKey: "modificationCounter")
```
Should managedObjectContext being nil for this be an assertionFailure, so we see any code mistakes earlier?

darinkrauss replied:
Sure. I'll use `managedObjectContext!` instead of `managedObjectContext?`.

darinkrauss replied:
Done.

ps2 commented on LoopKit/CarbKit/CarbStore.swift:
```diff
@@ -945,3 +843,99 @@ extension CarbStore {
+        
+        public typealias RawValue = [String: Any]
+        
+        public var deletedModificationCounter: Int64
```
QueryAnchors should be opaque.

ps2 replied:
Also, I don't think specifying the bit width is necessary here; we don't have any 32 bit platforms we're targeting.

darinkrauss replied:
How exactly would you like me to make them opaque?

Also, I can change the `Int64` to just `Int`.

darinkrauss replied:
Actually, it is an `Int64` because that matches the type of the field in each Core Data object. I can change the type of the field Core Data Swift object to be `Int`, but the actual data model requires a bit width so there'd have to be some type conversion somewhere along the way.

I will leave it as-is unless you feel strongly about it. If you want it changed, please let me know at what place you'd like the type conversion.

ps2 replied:
By opaque, I mean that `deletedModificationCounter` and `storedModificationCounter` should not be public.

darinkrauss replied:
Done.

ps2 commented on LoopKit/InsulinKit/DoseStore.swift:
```diff
@@ -1431,3 +1383,118 @@ extension DoseStore {
+
+    private var dosePredicate: NSPredicate {
+        return NSCompoundPredicate(orPredicateWithSubpredicates: [
+            NSPredicate(format: "type IN %@", PumpEventType.doseTypes.map { $0.rawValue }),
```
In what situations do we expect to have types other than these stored in the database?

darinkrauss replied:
In a previous conversation you suggested I break apart uploading doses from other pump events. That's what happens here. You'd have a better idea of what non-dose pump events are stored in the database.

ps2 commented on LoopKit/RemoteDataService.swift:
```diff
@@ -6,31 +6,92 @@
+    var carbDataLimit: Int { get }
+
+    /**
+     Synchronize carb data with the remote data service.
```
Nit: This seems more like uploading (one way) than synchronizing (two way).

darinkrauss replied:
Will be renamed to `upload...`.

darinkrauss replied:
Done.

ps2 commented on LoopKit/RemoteDataService.swift:
```diff
@@ -6,31 +6,92 @@
+
+public extension RemoteDataService {
+
+    var carbDataLimit: Int { return Int.max }
```
Better to use an optional set to nil, than Int.max, to denote no limit.

darinkrauss replied:
Will change.

darinkrauss replied:
Done.

ps2 commented on LoopKit/SettingsStore.swift:
```diff
@@ -0,0 +1,166 @@
+    
+    private let userDefaults: UserDefaults?
+    
+    private var settings: [Int64: StoredSettings]
```
No need for bit width specification.  We'll eventually need to persist this, of course, but that is covered in https://tidepool.atlassian.net/browse/LOOP-884, looks like.

darinkrauss replied:
Will change `Int64` in all anchors and related to `Int`.

darinkrauss replied:
Referencing above comment:

"Actually, it is an Int64 because that matches the type of the field in each Core Data object. I can change the type of the field Core Data Swift object to be Int, but the actual data model requires a bit width so there'd have to be some type conversion somewhere along the way.

I will leave it as-is unless you feel strongly about it. If you want it changed, please let me know at what place you'd like the type conversion."

ps2 commented on LoopKit/StatusStore.swift:
```diff
@@ -0,0 +1,188 @@
+    
+}
+
+public class StatusStore {
```
nit: "Status" is pretty vague. Would `LoopStatus` or `DosingDecisionData` be better?

darinkrauss replied:
Which would you prefer?

ps2 replied:
I think `DosingDecision` as the data object, and `DosingDecisionStore` as the store name could help scope the content better to help prevent this from becoming more of a dumping ground.

ps2 replied:
I'm open to other suggestions.  I just see "status" as something that, if not scoped, would be interpreted to mean *any* status of the entire Loop app, which does change more often than just dosing decision points.

darinkrauss replied:
Done.

ps2 commented on LoopKitTests/CarbStoreTests.swift:
```diff
@@ -5,443 +5,446 @@
-import CoreData
-@testable import LoopKit
+// TODO: Determine how to make these tests valid with new pull synchronization since
+// these tests tightly tie together both persistence and synchronization tests at once
```
Yeah, the actions of storing and uploading have been split apart (yay!), but this means that we have a significantly different set of interactions with the store, and probably new tests are warranted.

darinkrauss replied:
Done. Updated these tests and added many new tests for all of the new query anchors and store query code.

ps2 commented on LoopKitTests/TemporaryScheduleOverrideHistoryTests.swift:
```diff
@@ -14,11 +14,14 @@ final class TemporaryScheduleOverrideHistoryTests: XCTestCase {
+            RepeatingScheduleValue(startTime: .hours(6), value: 1.4),
+            RepeatingScheduleValue(startTime: .hours(20), value: 1.0)
+        ],
+        timeZone: Calendar.current.timeZone
```
This will have different values depending on where the test is run. Will the test fail in some zones?

darinkrauss replied:
Well, without this, it fails on my machine. Is there some magic I have to do to get this to work without the `timeZone` specified?

ps2 replied:
We can specify timezone; it just should be test-machine independent. So maybe just create create a fixed offset timezone of -0600, or UTC

darinkrauss replied:
Ah, the root problem is that code a few lines above, the `referenceDate` uses the current calendar, which uses the test machine's time zone. I'll update both to use GMT.

darinkrauss replied:
Done.

ps2 reviewed (changes requested):
Really happy to see this come together!  Looking really good. A few small changes, like making anchors opaque, and some nits.

darinkrauss committed:
Updates from PR review

darinkrauss committed:
Fix bug with settings and status store modification counters and limit

darinkrauss committed:
Fix DeviceDataManager in LoopKit Example

darinkrauss committed:
Store query updates and tests

- Check for non-zero limit for store queries
- Fix DoseStore queries for dose and non-dose data
- Update SettingsStore and StatusStore to be test friendly
- Fix returned query anchor for SettingsStore and StatusStore with limit
- Implement HKHealthStoreMock.deleteObjects function
- Reinstate CarbStore persistence tests
- Add various for store persistence, query anchors, and queries

darinkrauss commented:
@ps2 A number of changes to address most of your comments/questions. There are few remaining questions for you to answer, but as far as I'm concerned, I'm happy with it. 

Please review. If you approve I will rebase everything on top of `dev` (or merge) so we can get a final approval and merge.

ps2 reviewed (changes requested):
Looking good. Only remaining items are making the anchor opaque, and renaming the StatusStore to something more specific.

darinkrauss committed:
Rename Status* to DosingDecision* per PR review

darinkrauss committed:
Update modification counters to internal access per PR review

darinkrauss committed:
Use consistent calendar with GMT time zone for tests

ps2 reviewed (approved):
LGTM!

darinkrauss commented:
Closing in favor of final PR https://github.com/tidepool-org/LoopKit/pull/46.
